### ISSUE TITLE
OSX Rundown

### ISSUE THREAD

**Issue by sakkaku** [luke-jr/bfgminer#195]
**Status:** OpenCL kernel issue/detection?
**Status:** OpenCL header compile problem - **resolved in git**
**Status:** Mac endian.h missing - **resolved in git**
**Status:** `--enable-cpumining` doesn't link

Greetings.  I had an OSX machine I had lying around and figured I would test it and point out any issues.  The initial portion of this deals with building the latest tagged version in the repository.  With the last segment covering issues building against HEAD.
#### Requirements / Building

First up the method.  I chose to make a brew formula since it doesn't detect the system's curl. 

```
checking for LIBCURL... no
configure: error: Missing required libcurl dev >= 7.18.2
```

HomeBrew has a curl package but it doesn't install globally to avoid conflicts.  So I just made a [formula](https://gist.github.com/c0a4aa8c7c3cba8a555c) for [homebrew](http://mxcl.github.com/homebrew/) to get around that small issue with a forced breakpoint to test the binary incase it works.

Without the formula you can install all the dependencies with

```
brew install autoconf automake pkg-config libtool jansson yasm libusb
```

automake was missing from the dependencies list in the readme causing it to error out on aclocal being missing.  Also the AMD APP SDK link is a 404 now so you may need to fix that.
#### CPU Mining

With cpumining enabled there is a [linker error that I don't know how to fix](https://gist.github.com/743354baace43ffd92c3/46c2b65d2ce9f58767226ed5e57f3170e5140ca2).  This happens when building with both clang and gcc-llvm.
#### GPUs

GPUs don't seem to work.  I only have a shitty nvidia 9400 on this mac mini so I wasn't expecting much anyway.  Upon connecting to a pool it throws:

```
[2012-12-16 16:17:31] Probing for an alive pool
[2012-12-16 16:17:32] Long-polling activated for http://mine.btcguild.com:8332/LP/
[2012-12-16 16:17:36] Pool 0 http://mine.btcguild.com:8332 alive
[2012-12-16 16:17:36] Error: Enqueueing kernel onto command queue. (clEnqueueNDRangeKernel)
[2012-12-16 16:17:36] Thread 0 failure, exiting
[2012-12-16 16:17:36] Disabling extra threads due to dynamic mode.
[2012-12-16 16:17:36] Tune dynamic intensity with --gpu-dyninterval
[2012-12-16 16:17:36] Error: Enqueueing kernel onto command queue. (clEnqueueNDRangeKernel)
[2012-12-16 16:17:36] Thread 1 failure, exiting
[2012-12-16 16:18:39] Device 0 idle for more than 60 seconds, GPU 0 declared SICK!
```

Then it loops those messages unless you explicitly disable the GPU.  Since most OSX systems don't have a GPU worthwhile this isn't really necessary IMO.
#### HEAD Issues

When building against head there is an issue with the endian detection.

```
In file included from miner.c:55:
./miner.h:155:12: fatal error: 'endian.h' file not found
   #include <endian.h>
```

`miner.h` was fairly weird as it handles it fine around 122-127 with:

```
#if HAVE_BYTESWAP_H
  #include <byteswap.h>
#elif defined(USE_SYS_ENDIAN_H)
  #include <sys/endian.h>
#elif defined(__APPLE__)
  #include <libkern/OSByteOrder.h>
```

then fails starting at 146 with it falling back on endian.h by default. 

```
#ifdef WIN32
  #ifndef __LITTLE_ENDIAN
    #define __LITTLE_ENDIAN 1234
    #define __BIG_ENDIAN    4321
  #endif
  #ifndef __BYTE_ORDER
    #define __BYTE_ORDER __LITTLE_ENDIAN
  #endif
#else
  #include <endian.h>
```

Also CL/cl.h produces a ton of errors.  Starting with:

```
CC       bfgminer-miner.o
In file included from miner.h:22,
             from util.c:51:
CL/cl.h: In function ‘clCreateSubDevices’:
CL/cl.h:595: error: expected declaration specifiers before ‘CL_API_SUFFIX__VERSION_1_2’
```

or Clang

```
CC       bfgminer-miner.o
In file included from util.c:51:
In file included from ./miner.h:22:
    ./CL/cl.h:595:80: error: expected function body after function declarator
               cl_uint *                            /* num_devices_ret */) CL_API_SUFFIX__VERSION_1_2;
```

**Comment by luke-jr** [comment]
If it doesn't install cURL globally, where does it install it? The purpose of autoconf is to detect things like this, so it'd be nice to support it.

Regarding CPU mining, I'm guessing there's an ABI issue involving the leading underscore. CalcSha256_x64_sse4 is defined in x86_64/sha256_sse4_amd64.asm. Probably there's some GNUism we need to tell GCC not to use the leading underscore for the assembly functions.

Regarding OpenCL mining, did you try any of the GPU options, such as other kernels?

What do you consider "latest tagged" to be? There haven't been any changes to endian or OpenCL handling since 2.10.0.

**Comment by sakkaku** [comment]
> Regarding OpenCL mining, did you try any of the GPU options, such as other kernels?

Doh.  Yeah I went back and diablo worked fine.  I assume the others are patched to take advantage of ATI cards hence the crashing.

> If it doesn't install cURL globally, where does it install it? The purpose of autoconf is to detect things like this, so it'd be nice to support it.

haha. welcome to OSX.  You will probably gag at this.

```
mithril:Formula sakkaku$ locate curl.h
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.7.sdk/usr/include/curl/curl.h
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.8.sdk/usr/include/curl/curl.h
```

and

```
/usr/bin/curl
/usr/bin/curl-config
/usr/lib/libcurl.3.dylib
/usr/lib/libcurl.4.dylib
/usr/lib/libcurl.dylib
```

the brew formula fixes this easily

> What do you consider "latest tagged" to be? There haven't been any changes to endian or OpenCL handling since 2.10.0.

Doh, I was pulling against 2.3.1-2 as I didn't scroll down far enough to see the tag formats had changed >_<

Changing to 2.10 and 2.9.5 results in it stalling at

```
bash-3.2$ make
/usr/bin/make  all-recursive
Making all in lib
/usr/bin/make  all-recursive
make[4]: Nothing to be done for `all-am'.
Making all in ccan
make[2]: Nothing to be done for `all'.
Making all in libblkmaker
make[2]: *** No rule to make target `all'.  Stop.
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

of course libblkmaker/ is empty >_<  Anyway a git clone to fix that and we are back to CL/cl.h errors and endian issue.

**Comment by luke-jr** [comment]
I think they shouldn't patch unless an AMD card is detected, so maybe something to check into still... Can you post the OpenCL info for your card from --debug?

Looks like you didn't re-run autogen.sh after changing to newer tags. That uses git to populate libblkmaker. ;)

**Comment by sakkaku** [comment]
Ok to fix the endian issue add a line to detect OSX and do nothing.

```
#ifdef WIN32
  #ifndef __LITTLE_ENDIAN
    #define __LITTLE_ENDIAN 1234
    #define __BIG_ENDIAN    4321
  #endif
  #ifndef __BYTE_ORDER
    #define __BYTE_ORDER __LITTLE_ENDIAN
  #endif
#elif defined(__APPLE__)
  // do nothing as <libkern/OSByteOrder.h> has been included above
#else
  #include <endian.h>
#endif
```

**Comment by sakkaku** [comment]
> Looks like you didn't re-run autogen.sh after changing to newer tags. That uses git to populate libblkmaker. ;)

It errors out because the zip files from the tags don't include the .git directory >_<

**Comment by luke-jr** [comment]
Yeah, don't grab zip files from GitHub... ;)
Official source packages are at http://luke.dashjr.org/programs/bitcoin/files/bfgminer/ (and have libblkmaker pre-populated)

Can you test this code:

``` C
#if !(defined(__BYTE_ORDER) && defined(__LITTLE_ENDIAN) && defined(__BIG_ENDIAN))
  #ifdef WIN32
    #ifndef __LITTLE_ENDIAN
      #define __LITTLE_ENDIAN 1234
      #define __BIG_ENDIAN    4321
    #endif
    #ifndef __BYTE_ORDER
      #define __BYTE_ORDER __LITTLE_ENDIAN
    #endif
  #else
    #include <endian.h>
  #endif
#endif
```

**Comment by sakkaku** [comment]
Hmm that causes the file to be included still.  Which means those symbols aren't added >_<

I'm gonna go jump off a cliff now and do some more research.

**Comment by sakkaku** [comment]
This works.  For some reason `__LITTLE_ENDIAN__` was defined but `__BIG_ENDIAN__` was not.

```
#include <stdio.h>
#include <machine/endian.h>

#ifdef BYTE_ORDER
  #warning "Byte Order is Valid"
#endif

#ifdef BIG_ENDIAN
  #warning "Big Endian is Valid"
#endif

#ifdef LITTLE_ENDIAN
  #warning "Little Endian is Valid"
#endif

#define __LITTLE_ENDIAN LITTLE_ENDIAN
#define __BIG_ENDIAN BIG_ENDIAN
#define __BYTE_ORDER BYTE_ORDER

int main(void)
{
    printf("%i, %i, %i\n", __LITTLE_ENDIAN, __BIG_ENDIAN, __BYTE_ORDER);
}
```

Yay \o/

```
mithril:Desktop sakkaku$ gcc test.cpp 
test.cpp:5:4: warning: #warning "Byte Order is Valid"
test.cpp:9:4: warning: #warning "Big Endian is Valid"
test.cpp:13:4: warning: #warning "Little Endian is Valid"

mithril:Desktop sakkaku$ ./a.out 
1234, 4321, 1234
```

Thus this should work:

```
#if defined(WIN32)
  #ifndef __LITTLE_ENDIAN
    #define __LITTLE_ENDIAN 1234
    #define __BIG_ENDIAN    4321
  #endif
  #ifndef __BYTE_ORDER
    #define __BYTE_ORDER __LITTLE_ENDIAN
  #endif
#elif defined(__APPLE__)
  #include <machine/endian.h>
  #define __LITTLE_ENDIAN LITTLE_ENDIAN
  #define __BIG_ENDIAN BIG_ENDIAN
  #define __BYTE_ORDER BYTE_ORDER
#else
  #include <endian.h>
#endif
```

**Comment by sakkaku** [comment]
Is there a specific reason for CL/ being included with the project?

I would assume that Linux and OSX would have their own headers (though some it is of course OpenCL/cl.h on apple and CL/cl.h on other platforms).

Also looks like some copying of the header to `driver-opencl.c` and `ocl.c`.  Either that or I can't read tonight (which is probable).

**Comment by luke-jr** [comment]
CL/ is included because it is a standard interface and the headers are not provided by operating systems (at least not Linux). AMD and nVidia have their own OpenCL headers, but BFGMiner is portable at runtime by dynamically linking to any compliant OpenCL implementation. If I compiled releases against AMD's headers/library, BFGMiner might not run without the same environment, and the binary might include AMD code in violation of the GPL terms. The current arrangement is GPL-compliant and does not use any non-free software to build.

**Comment by sakkaku** [comment]
https://gist.github.com/6533a4e260cea8fbd4e3/

Any idea what is going on here?

Looking at CL/cl.h and OpenCL/cl.h it looks like the stuff that errors isn't in the OS provided headers....

clCreateSubDevices, clRetainDevice, clReleaseDevice, clCreateImage, clCreateProgramWithBuiltInKernels, etc.  All are OpenCL 1.2 specifications.

**Comment by sakkaku** [comment]
[Oooh.](https://github.com/luke-jr/bfgminer/blob/bfgminer/CL/cl_platform.h)  CL_API_SUFFIX__VERSION_1_2 is not defined under the `__APPLE__` conditional on line 48.

Here is the section of the header in the 10.8 API:

```
#ifdef __APPLE__
    #define CL_EXTENSION_WEAK_LINK       __attribute__((weak_import))
    #define CL_API_SUFFIX__VERSION_1_0                  AVAILABLE_MAC_OS_X_VERSION_10_6_AND_LATER
    #define CL_EXT_SUFFIX__VERSION_1_0                  CL_EXTENSION_WEAK_LINK AVAILABLE_MAC_OS_X_VERSION_10_6_AND_LATER
    #define CL_API_SUFFIX__VERSION_1_1                  AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER
    #define GCL_API_SUFFIX__VERSION_1_1                 AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER
    #define CL_EXT_SUFFIX__VERSION_1_1                  CL_EXTENSION_WEAK_LINK AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER
    #define CL_EXT_SUFFIX__VERSION_1_0_DEPRECATED       CL_EXTENSION_WEAK_LINK AVAILABLE_MAC_OS_X_VERSION_10_6_AND_LATER_BUT_DEPRECATED_IN_MAC_OS_X_VERSION_10_7
    #define CL_USE_DEPRECATED_OPENCL_1_1_APIS           1

    #ifdef AVAILABLE_MAC_OS_X_VERSION_10_8_AND_LATER
        #define CL_API_SUFFIX__VERSION_1_2              AVAILABLE_MAC_OS_X_VERSION_10_8_AND_LATER
        #define GCL_API_SUFFIX__VERSION_1_2             AVAILABLE_MAC_OS_X_VERSION_10_8_AND_LATER
        #define CL_EXT_SUFFIX__VERSION_1_2              CL_EXTENSION_WEAK_LINK AVAILABLE_MAC_OS_X_VERSION_10_8_AND_LATER
        #define CL_EXT_SUFFIX__VERSION_1_1_DEPRECATED   CL_EXTENSION_WEAK_LINK AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER_BUT_DEPRECATED_IN_MAC_OS_X_VERSION_10_8
    #else
        #warning  This path should never happen outside of internal operating system development.  AvailabilityMacros do not function correctly here!
        #define CL_API_SUFFIX__VERSION_1_2              AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER
        #define GCL_API_SUFFIX__VERSION_1_2             AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER
        #define CL_EXT_SUFFIX__VERSION_1_2              CL_EXTENSION_WEAK_LINK AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER
        #define CL_EXT_SUFFIX__VERSION_1_1_DEPRECATED   CL_EXTENSION_WEAK_LINK AVAILABLE_MAC_OS_X_VERSION_10_7_AND_LATER
    #endif
#else
    #define CL_API_SUFFIX__VERSION_1_0
    #define CL_API_SUFFIX__VERSION_1_1
    #define CL_API_SUFFIX__VERSION_1_2
    #define CL_EXTENSION_WEAK_LINK
#endif
```

**Comment by luke-jr** [comment]
Hmm, does the OpenCL problem solve if you update the CL headers from http://www.khronos.org/registry/cl/ ?

**Comment by sakkaku** [comment]
Wow.  That was a little hacky.  In `cl.h` you have to change

```
#ifdef __APPLE__
  #include <OpenCL/cl_platform.h>
#else
  #include "CL/cl_platform.h"
#endif
```

to

```
#include "cl_platform.h"
```

It was still erroring on `CL_EXT_PREFIX__VERSION_1_1_DEPRECATED` until I caught that. >_<

Now we are back to the errors I was getting before I forcefully removed the copied headers from `driver-opencl` and `ocl` to use the system's opencl (of which that was a fairly bad idea as it didn't work heh).

```
*: error: redefinition of '*' as different kind of symbol
```

or if you enjoy GCCs error syntax.

```
ocl.c:41: error: ‘clGetPlatformIDs’ redeclared as different kind of symbol
CL/cl.h:569: error: previous declaration of ‘clGetPlatformIDs’ was here
```

https://gist.github.com/743354baace43ffd92c3/fbca133d2b648ce3cd84050a96e74fedb10e08a4

I assume the redefinitions inside [driver_opencl.c](https://github.com/sakkaku/bfgminer/blob/bfgminer/driver-opencl.c) and [ocl.c](https://github.com/sakkaku/bfgminer/blob/bfgminer/ocl.c) will need to be updated to work with the newer opencl headers.  It looks like you could throw both sets into one file with a preprocessor directive to omit `extern` for one set and help ease maintenance.

**Comment by luke-jr** [comment]
Does the OpenCL problem solve if you update the CL headers from http://www.khronos.org/registry/cl/ ?

Are the __BYTE_ORDER macros defined if you include <sys/param.h>? Can you post the output of this:

```
cpp -dM <<< '#include <sys/param.h>' | grep ENDIAN
```

**Comment by sakkaku** [comment]
> Does the OpenCL problem solve if you update the CL headers from http://www.khronos.org/registry/cl/ ?

No.  Specifically it causes a different set of problems which I think is do to the mismatch between the new headers and definitions in [driver_opencl.c](https://github.com/sakkaku/bfgminer/blob/bfgminer/driver-opencl.c) and [ocl.c](https://github.com/sakkaku/bfgminer/blob/bfgminer/ocl.c).  [The errors are posted here](https://gist.github.com/743354baace43ffd92c3/fbca133d2b648ce3cd84050a96e74fedb10e08a4).

```
 mithril:Desktop sakkaku$ cpp -dM <<< '#include <sys/param.h>' | grep ENDIAN
 #define _I386__ENDIAN_H_ 
 #define _SYS__ENDIAN_H_ 
 #define __DARWIN_PDP_ENDIAN 3412
 #define __DARWIN_BYTE_ORDER __DARWIN_LITTLE_ENDIAN
 #define __DARWIN_BIG_ENDIAN 4321
 #define PDP_ENDIAN __DARWIN_PDP_ENDIAN
 #define BIG_ENDIAN __DARWIN_BIG_ENDIAN
 #define __DARWIN_LITTLE_ENDIAN 1234
 #define __LITTLE_ENDIAN__ 1
 #define _BSD_MACHINE_ENDIAN_H_ 
 #define LITTLE_ENDIAN __DARWIN_LITTLE_ENDIAN

mithril:Desktop sakkaku$ cpp -dM <<< '#include <sys/param.h>' | grep BYTE_ORDER
#define __DARWIN_BYTE_ORDER __DARWIN_LITTLE_ENDIAN
#define BYTE_ORDER __DARWIN_BYTE_ORDER
```

**Comment by luke-jr** [comment]
Please pull and check that the new endian/byteswap checks work on Mac.

**Comment by luke-jr** [comment]
Also please report on the updated CL/\* in git too.

**Comment by sakkaku** [comment]
```
  CCLD     bfgminer
Undefined symbols for architecture x86_64:
  "_OSSwapInt32", referenced from:
      _share_diff in bfgminer-miner.o
      _submit_work_thread in bfgminer-miner.o
      _gen_stratum_work in bfgminer-miner.o
      _roll_work in bfgminer-miner.o
      _work_decode in bfgminer-miner.o
      _calc_diff in bfgminer-miner.o
      _calc_midstate in bfgminer-miner.o
      ...
  "_OSSwapInt64", referenced from:
      _share_diff in bfgminer-miner.o
      _gen_stratum_work in bfgminer-miner.o
      _calc_diff in bfgminer-miner.o
      _scrypt_outputhash in bfgminer-scrypt.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bfgminer] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

Doesn't work >_<.  Looks like it isn't linking to the byteswap functions.

The OpenGL stuff seems to have worked as it compiles if I use the miner.h header I had.

**Comment by luke-jr** [comment]
Looks like we're getting closer, though! Can you pastebin your config.log ?

**Comment by sakkaku** [comment]
Posting that into gist almost crashed my browser...  I'll just post it to dropbox, [config.log](http://dl.dropbox.com/u/155051/config.log).

**Comment by luke-jr** [comment]
Hmm, this is odd. When configure links its test, it succeeds, but then not when the build system does it.
Please pastebin the output of: make AM_V_CCLD=''

**Comment by sakkaku** [comment]
> Please pastebin the output of: make AM_V_CCLD=''

https://gist.github.com/3efe5ef1d8df348b701e

Doesn't seem to be anything diff

**Comment by luke-jr** [comment]
brew: superenv removed: -g -O2

Can you stop it from removing -O2? :)

**Comment by sakkaku** [comment]
> Can you stop it from removing -O2? :)

Maybe.  A lot of stuff is lolundocumented.  I just recently found the `--use-llvm` and `--use-clang` flags so >_<

**Comment by luke-jr** [comment]
How about trying to build it without brew again?

```
./configure --enable-scrypt 'PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig:/usr/local/lib/pkgconfig:/usr/local/Library/ENV/pkgconfig/mountain_lion'
```

**Comment by sakkaku** [comment]
Same errors outside of brew even though that configure line worked.

https://gist.github.com/3efe5ef1d8df348b701e/

`-O2` is an optimizer flag and not too important.  Like omitting the radiator in a car, or never changing the oil; it still sorta runs right?

**Comment by luke-jr** [comment]
Can you post the config.log from that one? Looks different...

I am suspecting -O2 related in case of something weird going on with the function being inlined. Without optimization, inlines are ignored...

**Comment by sakkaku** [comment]
http://dl.dropbox.com/u/155051/config.log.2

**Comment by luke-jr** [comment]
How about config.h?

**Comment by luke-jr** [comment]
First, try the new commits I just pushed :)

**Comment by sakkaku** [comment]
Works \o/

```
mithril:bfgminer sakkaku$ ./bfgminer --debug
 [2012-12-18 03:08:34] ft232r_scan: Found 05ac:8006 - not a ft232r                    
 [2012-12-18 03:08:34] ft232r_scan: Found 05ac:8006 - not a ft232r                    
 [2012-12-18 03:08:34] ft232r_scan: Found 05ac:8005 - not a ft232r                    
 [2012-12-18 03:08:34] ft232r_scan: Found 05ac:8005 - not a ft232r                    
 [2012-12-18 03:08:34] ft232r_scan: Found 0557:7000 - not a ft232r                    
 [2012-12-18 03:08:34] ft232r_scan: Found 05ac:8242 - not a ft232r                    
 [2012-12-18 03:08:34] ft232r_scan: Found 04d9:2013 - not a ft232r                    
 [2012-12-18 03:08:34] ft232r_scan: Found 046d:c066 - not a ft232r                    
All devices disabled, cannot mine!
```

But it looks like it doesn't detect the GPU (the one above where it did detect was from the really old builds though so it probably isn't a recent regression).

Any was to get more debugging info out?  Also [config.h](https://dl.dropbox.com/u/155051/config.h).

For CPU it looks like this error might be relevant.

```
Making all in x86_64
/usr/local/bin/yasm -f elf64 sha256_xmm_amd64.asm
/usr/local/bin/yasm -f elf64 sha256_sse4_amd64.asm
  AR       libx8664.a
ranlib: warning for library: libx8664.a the table of contents is empty (no object file members in the library define global symbols)
```

Does this get omitted on other systems?  That might explain those linker errors with `--enable-cpumining`.

Then the linker errors again to avoid scrolling.

```
ld: warning: ignoring file x86_64/libx8664.a, file was built for archive which is not the architecture being linked (x86_64): x86_64/libx8664.a
Undefined symbols for architecture x86_64:
  "_CalcSha256_x64_sse4", referenced from:
      _scanhash_sse4_64 in bfgminer-sha256_sse4_amd64.o
  "_sha256_sse2_64_new", referenced from:
      _scanhash_sse2_64 in bfgminer-sha256_sse2_amd64.o
ld: symbol(s) not found for architecture x86_64
collect2: ld returned 1 exit status
make[2]: *** [bfgminer] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

**Comment by luke-jr** [comment]
The current OpenCL driver is looking for "libOpenCL.so" specifically. I presume that probably needs to be something different on Mac, but what? :)

**Comment by luke-jr** [comment]
Does this patch help the assembly? http://codepad.org/YYm9nqoG

**Comment by sakkaku** [comment]
> I presume that probably needs to be something different on Mac, but what? :)

```
gcc -Wl,-t test.cpp -framework OpenCL
/usr/lib/crt1.10.6.o
/var/folders/9b/ccbl3jg95_9dtw060sqb5hdm0000gn/T//cci2dZpU.o
/usr/lib/libSystem.dylib
/System/Library/Frameworks//OpenCL.framework/OpenCL
.....

mithril:OpenCL.framework sakkaku$ nm /System/Library/Frameworks//OpenCL.framework/OpenCL | grep clCreateSubDevice
000000000001e036 T _clCreateSubDevices
```

I assume that would work >_<.

> Does this patch help the assembly? http://codepad.org/YYm9nqoG

Didn't seem to.  Same errors.

**Comment by luke-jr** [comment]
driver-opencl.c has a line:

``` C
void *cl = dlopen("libOpenCL.so", RTLD_LAZY);
```

Try changing it to:

``` C
void *cl = dlopen("/System/Library/Frameworks/OpenCL.framework/Versions/Current/OpenCL", RTLD_LAZY);
```

Regarding asm, did you re-run autogen.sh?

**Comment by sakkaku** [comment]
> Regarding asm, did you re-run autogen.sh?

Looks like I forgot the make clean >_<

```
Making all in x86_64
/usr/local/bin/yasm -o sha256_xmm_amd64.o -m amd64 -f macho64 sha256_xmm_amd64.asm
sha256_xmm_amd64.asm:294: error: undefined symbol `_sha256_consts_m128i' (first use)
sha256_xmm_amd64.asm:294: error:  (Each undefined symbol is reported only once.)
make[2]: *** [sha256_xmm_amd64.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

> Try changing it to ...

That worked.  Unfortunately all 4 kernels don't run (heh).

```
[2012-12-18 13:13:27] Error -54: Enqueueing kernel onto command queue. (clEnqueueNDRangeKernel)
[2012-12-18 13:13:27] OCL 0 failure, disabling!
[2012-12-18 13:13:27] Thread 0 being disabled
[2012-12-18 13:13:30] Stratum from pool 0 requested work restart
[2012-12-18 13:13:39] Attempting to reinitialize OCL 0
[2012-12-18 13:13:39] Thread 0 being re-enabled
```

Is it possible it is trying to use the CPU as an OpenCL device?  I know OSX has support for that out of the box.

**Comment by luke-jr** [comment]
BFGMiner's OpenCL driver specifically asks for only GPUs. You can try the -d? option to get a list of devices. Adding --debug might get more info too. (Please pastebin it)

**Comment by sakkaku** [comment]
looks like the extern is sha256_consts_m128i

but this line(257) refers to a diff name.

```
  lea       rcx, [_sha256_consts_m128i wrt rip]
```

removing the leading underscore causes

```
/usr/local/bin/yasm -o sha256_xmm_amd64.o -m amd64 -f macho64 sha256_xmm_amd64.asm
sha256_xmm_amd64.asm:305: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
```

I am bad at ASM though >_<

> You can try the -d? option to get a list of devices.

```
mithril:bfgminer sakkaku$ ./bfgminer -d
 [2012-12-18 13:31:52] ./bfgminer: -d: requires an argument

mithril:bfgminer sakkaku$ ./bfgminer -n
 [2012-12-18 13:35:38] CL Platform 0 vendor: Apple                    
 [2012-12-18 13:35:38] CL Platform 0 name: Apple                    
 [2012-12-18 13:35:38] CL Platform 0 version: OpenCL 1.2 (Aug 24 2012 00:53:09)                    
 [2012-12-18 13:35:38] Platform 0 devices: 1                    
 [2012-12-18 13:35:38]  0   GeForce 9400                    
 [2012-12-18 13:35:38] Unable to load ati adl library                    
 [2012-12-18 13:35:38] 1 GPU devices max detected
```

**Comment by luke-jr** [comment]
Here's another attempt at a patch for asm: http://codepad.org/HIPCStvd

-d? had a literal questionmark. So, run exactly: ./bfgminer -D -T -d?

**Comment by sakkaku** [comment]
> Here's another attempt at a patch for asm: http://codepad.org/HIPCStvd

```
Making all in x86_64
/usr/local/bin/yasm -o sha256_xmm_amd64.o -m amd64 -f macho64 sha256_xmm_amd64.asm
sha256_xmm_amd64.asm:301: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
```

**Comment by luke-jr** [comment]
http://codepad.org/nnhbjn0V

**Comment by sakkaku** [comment]
> http://codepad.org/nnhbjn0V

```
Making all in x86_64
/usr/local/bin/yasm -o sha256_xmm_amd64.o -m amd64 -f macho64 sha256_xmm_amd64.asm
/usr/local/bin/yasm -o sha256_sse4_amd64.o -m amd64 -f macho64 sha256_sse4_amd64.asm
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
sha256_sse4_amd64.asm:220: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.
```

One down.

**Comment by luke-jr** [comment]
http://codepad.org/i2rZVR70

**Comment by sakkaku** [comment]
That compiles.

```
mithril:bfgminer sakkaku$ ./bfgminer --enable-cpu --algo auto
[2012-12-18 22:49:50] benchmarking all sha256 algorithms ...                    
[2012-12-18 22:49:50] "c"        : benchmarking algorithm ...                    
[2012-12-18 22:49:56] "c"        : algorithm runs at 0.79627 MH/s                    
[2012-12-18 22:49:56] "4way"     : benchmarking algorithm ...                    
[2012-12-18 22:50:02] "4way"     : algorithm runs at 0.63827 MH/s                    
[2012-12-18 22:50:02] "cryptopp" : benchmarking algorithm ...                    
[2012-12-18 22:50:08] "cryptopp" : algorithm runs at 0.72295 MH/s                    
[2012-12-18 22:50:08] "sse2_64"  : benchmarking algorithm ...                    
Segmentation fault: 11
```

Using any algorithm (via `--algo`) results in segfault shortly after starting.

[Here is the output.](https://gist.github.com/faff6f59b9eb390c4074)

[Also with OpenCL / Diablo.](https://gist.github.com/68791bee72b06cea6f98)

```
(gdb) run
Starting program: /Users/sakkaku/Downloads/bfgminer/bfgminer --enable-cpu --algo auto
Reading symbols for shared libraries +++++++................................................... done
 [2012-12-18 23:11:50] benchmarking all sha256 algorithms ...                    
 [2012-12-18 23:11:50] "c"        : benchmarking algorithm ...                    
 [2012-12-18 23:11:55] "c"        : algorithm runs at 0.80250 MH/s                    
 [2012-12-18 23:11:55] "4way"     : benchmarking algorithm ...                    
 [2012-12-18 23:12:02] "4way"     : algorithm runs at 0.62577 MH/s                    
 [2012-12-18 23:12:02] "cryptopp" : benchmarking algorithm ...                    
 [2012-12-18 23:12:08] "cryptopp" : algorithm runs at 0.64422 MH/s                    
 [2012-12-18 23:12:08] "sse2_64"  : benchmarking algorithm ...                    

Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: 13 at address: 0x0000000000000000
0x00000001000564ce in ..@5.SHA_CALC_W ()
(gdb) stacktrace
Undefined command: "stacktrace".  Try "help".
(gdb) backtrace
#0  0x00000001000564ce in ..@5.SHA_CALC_W ()
#1  0x00007fff5fbfee40 in ?? ()
#2  0x000000010004635f in bench_algo_stage3 (algo=1024) at driver-cpu.c:249
#3  0x00000001000464aa in bench_algo (best_rate=0x7fff5fbff158, best_algo=0x7fff5fbff154, algo=1606412032) at driver-cpu.c:553
#4  0x00000001000465cc in pick_fastest_algo [inlined] () at /Users/sakkaku/Downloads/bfgminer/driver-cpu.c:648
#5  0x00000001000465cc in set_algo (arg=0x7fff5fbff158 "?\032??\f???\001", algo=0x100074eb0) at driver-cpu.c:681
#6  0x0000000100055650 in parse_one (argc=0x7fff5fbff1e0, argv=0x7fff5fbff1e0, offset=0x7fff5fbff1e0, errlog=0x7fff5fbff1e0) at parse.c:110
#7  0x0000000100054c11 in opt_parse (argc=0x7fff5fbffa8c, argv=0x7fff5fbffb00, errlog=0x1000041c0 <applog_and_exit>) at opt.c:213
#8  0x0000000100013206 in main (argc=3, argv=0x7fff5fbffae0) at miner.c:7509
(gdb)


(gdb) run
Starting program: /Users/sakkaku/Downloads/bfgminer/bfgminer --enable-cpu --algo c
Reading symbols for shared libraries +++++++................................................... done
Reading symbols for shared libraries . done

Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: 13 at address: 0x0000000000000000
[Switching to process 97467 thread 0x6703]
0x00000001000564ce in ..@5.SHA_CALC_W ()
(gdb) backtrace
#0  0x00000001000564ce in ..@5.SHA_CALC_W ()
#1  0x0000000100342340 in ?? ()
#2  0x0000000100045f23 in cpu_scanhash (thr=0x103b2dde0, work=0x0, max_nonce=4357021152) at driver-cpu.c:825
#3  0x000000010000d31f in miner_thread (userdata=0x10100faf0) at miner.c:6300
#4  0x00007fff8f710742 in _pthread_start ()
#5  0x00007fff8f6fd181 in thread_start ()
(gdb)
```

Looks like fun!

**Comment by luke-jr** [comment]
Dead end on the CPU mining I guess, I wouldn't have the foggiest idea how to debug the assembly...

For OpenCL, try playing with the --intensity, --vectors, and/or --worksize options?

**Comment by sakkaku** [comment]
> For OpenCL, try playing with the --intensity, --vectors, and/or --worksize options?

Doesn't seem to be doing anything... not that a Geforce 9400m does much anyway.

```
#define CL_INVALID_WORK_GROUP_SIZE                  -54
```

Looking at the headers and error, -54 seems to refer to the worksize.  Diablominer(loljava) works fine with a "local worksize" of 512.  I tried stuff along 2^n: 4, 8, 16, 32, 64, 128, 256 and 512.

[Might be relevant:](https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clEnqueueNDRangeKernel.html)
- `CL_INVALID_WORK_GROUP_SIZE` if local_work_size is specified and number of work-items specified by global_work_size is not evenly divisable by size of work-group given by local_work_size or does not match the work-group size specified for kernel using the `__attribute__ ((reqd_work_group_size(X, Y, Z)))` qualifier in program source.
- `CL_INVALID_WORK_GROUP_SIZE` if local_work_size is specified and the total number of work-items in the work-group computed as `local_work_size[0] *... local_work_size[work_dim - 1]` is greater than the value specified by `CL_DEVICE_MAX_WORK_GROUP_SIZE` in the table of OpenCL Device Queries for clGetDeviceInfo.
- `CL_INVALID_WORK_GROUP_SIZE` if local_work_size is `NULL` and the `__attribute__ ((reqd_work_group_size(X, Y, Z)))` qualifier is used to declare the work-group size for kernel in the program source.

**Comment by peterhost** [comment]
Just for information : I managed to get it working on `OSX.8` with a `NVIDIA GeForce GTX 680MX 2048 MB`, by applying the parts of the last patch in this thread which were not yet merged into `master` and by indicating by hand the `PKG_CONFIG_PATH` path to homebrew's libcurl at `configure` time, like you indicated.

Just playing around. It works, but for the moment I haven't been able to reach the hashrates i'd expect from this GPU, even with `scrypt`,... without crashing the system ;)

```
bfgminer version 2.10.2 - Started: [2013-01-06 18:14:45] - [  0 days 00:00:43]
--------------------------------------------------------------------------------
5s:  1.8 avg:  3.8 u:  0.0 kh/s | A:1 R:0 S:0 HW:0 U:2.9/m
ST:2 DW:6 GW:5 LW:18 GF:0 NB:1 AS:0 RF:0 E:0.00 BW:  1.0kh/s
Connected to 127.0.0.1 diff 0 with stratum as user *******
Block: 7cdf60bdc6d3fe7c...  Diff:1.71M  Started: [18:14:45]  Best share: 1
--------------------------------------------------------------------------------
[P]ool management [G]PU management [S]ettings [D]isplay options [Q]uit
OCL 0:                |   3.9/  3.8/  0.0kh/s | A:1 R:0 HW:0 U:2.85/m
--------------------------------------------------------------------------------
```

NB: I'm using P2pool

**Comment by luke-jr** [comment]
Can you pastebin the diff you're using, more specifically?

Thanks

**Comment by sakkaku** [comment]
Probably the [patches to the assembly](https://github.com/luke-jr/bfgminer/issues/195#issuecomment-11517112) which didn't work anyway (by default the CPU isn't used or even support compiled for).  Master compiles and runs (aside from the errors I get at runtime) with and without the patch for me (specifically the patch is still needed for cpumining to be compiled).

If you want to test the assembly patch add `--enable-cpumining` to the configure line and pass `--enable-cpu` to bfgminer.

**Comment by luke-jr** [comment]
Does it not build without cpumining support? :/

**Comment by sakkaku** [comment]
It builds without.

**Comment by peterhost** [comment]
this one http://codepad.org/i2rZVR70 (you posted), ignoring what you pulled from that in your master branch and which is already in the code. Btw : i'm not a C dev, so my knowledge goes no further than using make/cmake.

I didn't try to enable cpu. Without it, it builds and works on my box with the most standard options. 

```
"intensity" : "d",
"vectors" : "1",
"worksize" : "256",
"kernel" : "scrypt",
"lookup-gap" : "0",
"thread-concurrency" : "0",
"shaders" : "0",
"gpu-engine" : "0-0",
"gpu-fan" : "0-0",
"gpu-memclock" : "0",
"gpu-memdiff" : "0",
"gpu-powertune" : "0",
"gpu-vddc" : "0.000",
"temp-cutoff" : "95",
"temp-overheat" : "0",
"temp-target" : "89",
"api-port" : "4028",
"expiry" : "120",
"expiry-lp" : "3600",
"gpu-dyninterval" : "7",
"gpu-platform" : "0",
"gpu-threads" : "2",
"log" : "5",
"queue" : "1",
"scan-time" : "60",
"scrypt" : true,
"skip-security-checks" : "0",
"submit-stale" : true,
"temp-hysteresis" : "3",
"shares" : "0",
"kernel-path" : "/usr/local/bin"
```

I get errors when I enable shaders. And as 'nvidia-smi' only ships with the *nix version of CUDA 5, i'm clueless (being a total dick as comes to C as I just said)

**Comment by sakkaku** [comment]
```
"gpu-dyninterval" : "7"
```

Try bumping this up.  On my windows desktop I usually use around 20-40 to get a good hash rate (old 5870).

In other news I managed to get GPU mining enabled on my rig after bumping the value up to 20+ (previously I was using command line options previously which might have been overwritten via the config file >_<)

```
Thread 0: 1.4 Mh/s Enabled ALIVE
```

Aww yeah, so fast you will cry.

**Comment by luke-jr** [comment]
@peterhost What problems did you have building without the patch? That one is supposed to only try to fix CPU mining...

**Comment by peterhost** [comment]
Sorry, it's just me being unclear. I had the usual macos "forgot libcurl was handled by homebrew" problem. This thread solved this issue. 
I stupidly applied the patch while not compiling with cpu support.

@sakkaku : thx.

**Comment by luke-jr** [comment]
So it works without the patch?

Will this configure line work as-is?

```
./configure PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig
```

**Comment by peterhost** [comment]
yes it does : http://pastebin.com/ks4sGgJQ

**Comment by luke-jr** [comment]
Ok, I added this to the README in git. Thanks

**Comment by fabulouspanda** [comment]
For those using MacPorts rather than Homebrew - I ran in to a problem with the jansson library not being loaded from the version installed that way, but downloading the latest version here http://www.digip.org/jansson/ and doing the standard ./configure make make install dropped the library in to /usr/local/lib/ and the make worked fine. Hope that helps someone. (all the other dependencies worked fine as installed by MacPorts).

Now if only I knew enough C to have it read the temperatures on a Mac!

**Comment by lbr** [comment]
RE: https://github.com/luke-jr/bfgminer/issues/195#issuecomment-11517759

Solved the same issue when compiling ckolivas-cgminer-b8f4fa4(ckolivas-cgminer-v2.11.3-12-gb8f4fa4.tar.gz) on MacOSX 10.6 by setting GCC CFLAGS to support SSE4.2, check http://en.gentoo-wiki.com/wiki/Hardware_CFLAGS#Determining_available_processor_features

I've used

export CC="gcc-4.8"
export CFLAGS="-O3 -pipe -march=core2 -mcx16 -msahf -mno-movbe -mno-aes -mno-pclmul -mno-popcnt -mno-abm -mno-lwp -mno-fma -mno-fma4 -mno-xop -mno-bmi -mno-bmi2 -mno-tbm -mno-avx -mno-avx2 -mno-sse4.2 -msse4.1 -mno-lzcnt -mno-rtm -mno-hle -mno-rdrnd -mno-f16c -mno-fsgsbase -mno-rdseed -mno-prfchw -mno-adx -mfxsr -mno-xsave -mno-xsaveopt --param l1-cache-size=32 --param l1-cache-line-size=64 --param l2-cache-size=3072 -mtune=core2 -fPIC -mmacosx-version-min=10.6.8"

export CXXFLAGS="${CFLAGS}"
./autogen.sh --enable-cpumining --enable-scrypt --build=x86_64-apple-darwin11.2.0

**Comment by prologic** [comment]
Is OpenCL working yet on OS X (Mac Book Air)?

I get this error from -D

```
 [2013-04-17 21:33:31] Started bfgminer 3.0.0
 [2013-04-17 21:33:31] DISPLAY not set, setting :0 just in case
 [2013-04-17 21:33:31] CL Platform 0 vendor: Apple
 [2013-04-17 21:33:31] CL Platform 0 name: Apple
 [2013-04-17 21:33:31] CL Platform 0 version: OpenCL 1.2 (Dec  4 2012 18:26:30)
 [2013-04-17 21:33:31] Error -1: Getting Device IDs (num)
```

I installed from head.

--JamesMills

**Comment by keenser** [comment]
here's my patch, which helped launch sse2 cpumining on macbook with core2duo https://docs.google.com/file/d/0B1U5lrPuFx6JaVVlcXlZck9CX2c/edit?usp=sharing

**Comment by deutrino** [comment]
I'm trying to build for OS X on 10.7 with homebrew using the following configure invocation:

```
/configure --enable-cpumining --enable-scrypt --disable-opencl --disable-bitforce --disable-icarus --disable-modminer --disable-x6500 --disable-ztex --without-libudev PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig:/usr/local/opt/jansson/lib/pkgconfig
```

When I do this on a clean tree (bfgminer-2.10.8), this is the result:

```
Making all in x86_64
/usr/local/bin/yasm -f elf64 sha256_xmm_amd64.asm
/usr/local/bin/yasm -f elf64 sha256_sse4_amd64.asm
  AR       libx8664.a
ranlib: warning for library: libx8664.a the table of contents is empty (no object file members in the library define global symbols)
  CC       bfgminer-miner.o
  CC       bfgminer-util.o
  CC       bfgminer-sha2.o
  CC       bfgminer-api.o
  CC       bfgminer-logging.o
  CC       bfgminer-driver-opencl.o
  CC       bfgminer-ocl.o
  CC       bfgminer-findnonce.o
  CC       bfgminer-adl.o
  CC       bfgminer-scrypt.o
  CC       bfgminer-sha256_generic.o
  CC       bfgminer-sha256_4way.o
  CC       bfgminer-sha256_via.o
  CC       bfgminer-sha256_cryptopp.o
  CC       bfgminer-sha256_sse2_amd64.o
  CC       bfgminer-sha256_sse4_amd64.o
  CC       bfgminer-sha256_sse2_i386.o
  CC       bfgminer-sha256_altivec_4way.o
  CC       bfgminer-driver-cpu.o
  CC       bfgminer-fpgautils.o
  CC       bfgminer-dynclock.o
  CCLD     bfgminer
ld: warning: ignoring file x86_64/libx8664.a, file was built for archive which is not the architecture being linked (x86_64)
Undefined symbols for architecture x86_64:
  "_sha256_sse2_64_new", referenced from:
      _scanhash_sse2_64 in bfgminer-sha256_sse2_amd64.o
  "_CalcSha256_x64_sse4", referenced from:
      _scanhash_sse4_64 in bfgminer-sha256_sse4_amd64.o
ld: symbol(s) not found for architecture x86_64
collect2: ld returned 1 exit status
make[2]: *** [bfgminer] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

Note that `-f elf64` for yasm is not correct on OS X, it should be `-f macho64` if I'm not mistaken.

I attempted to substitute `macho64` for `elf64` in `x86_64/Makefile.*`, but ran into other problems.

Apologies if this is a duplicate comment.

_Edit:_ I applied the patch at http://codepad.org/i2rZVR70 - same problem.

**Comment by lbr** [comment]
"Note that -f elf64 for yasm is not correct on OS X, it should be -f macho64 if I'm not mistaken.
I attempted to substitute macho64 for elf64 in x86_64/Makefile.*, but ran into other problems."

You are not mistaken, I've also replaced elf64 to maco64 and also ran into other issues, however I've solved them.
If I remember correctly there were some problems with __function instead of _function in asm files.
But I've compiled cgminer, if you want I can send you the source.

**Comment by luke-jr** [comment]
This was the furthest I got with CPU mining: http://codepad.org/i2rZVR70

**Comment by deutrino** [comment]
lbr: I saw your above comment about cgminer.  I'd be happy to try it, I'm about out of options for bfgminer as it won't even build for me with http://codepad.org/i2rZVR70 and some replacement of 'elf64' with 'macho64'.

(The primary reason I'm working on this is that there seems to be no way to connect cpuminer to a Stratum pool, mining_proxy is throwing 401s no matter what I do when minerd connects, and I have a flock of machines on which I want to do scrypt CPU mining.  A bit OT I know, but hit me up, anybody who may have got cpuminer working with Stratum pools.)

**Comment by lbr** [comment]
gordon, you can find my modified source of cgminer 2.11.3 here - 
https://lbr.id.lv/#cgminer_2_11_3_x64_portable_10_6_8
it compiles on my mac mini(core 2 duo + geforce 330m).
There is also pre-compiled 'portable' package with all the libs included.
It is already pre-configured and Makefiles are hard modified after(elf64 -> macho64).
Basically I've downloaded cgminer and started modifying it until it compiled successfully, then I've archived everything.
I suppose you can get the same revision(b8f4fa4) of cgminer from github and compare it with mine to make proper patch.

p.s. You can connect cpuminer to stratum pool using stratum proxy.

**Comment by deutrino** [comment]
lbr, thanks. I can't figure out how to send private messages on github so could you email me about this if interested, so I can keep the OT discussion off this bug? https://gist.github.com/gordon-morehouse/5b5b395daee41305abb5  Essentially, mining_proxy throws constant 401s back at cpuminer.  And I'd love to discuss your work, but off this bug.  :) Thanks!

**Comment by hashier** [comment]
I get the same linking errors:

curl from homebrew

git HEAD
./autogen.sh
patch -p1 < patch.file
./configure PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig --enable-cpumining --enable-scrypt
make
->
ld: warning: ignoring file x86_64/libx8664.a, file was built for archive which is not the architecture being linked (x86_64): x86_64/libx8664.a
Mountain Lion Mac Mini.

**Comment by ibd1279** [comment]
The patch modifies configure.ac, so you need to patch before you run autogen.sh.

**Comment by luke-jr** [comment]
I've tried to resolve some of the OS X CPUmining difficulties in latest git. If someone wants to test and report results, I'd appreciate it...

**Comment by fabulouspanda** [comment]
Thanks luke, that works fine with or without scrypt, i just git clone, 
./autogen.sh
./configure --enable-cpumining --enable-scrypt PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig:/usr/local/opt/jansson/lib/pkgconfig
make
./bfgminer -C -owith -umy -psettings

**Comment by luke-jr** [comment]
Really? I wasn't expecting it to be _that_ easy! CPU mining actually works and all?
For testing, you can use -o luke.dashjr.org/tmp/code/earlyshare2.json -O x (if it works, you will see a Rejected share pretty quick)

**Comment by fabulouspanda** [comment]
I tested it with my p2pool node and got accepted shares and now with your url rejected shares so it should be working fine! It is a lot slower than poolers cpu miner though…

**Comment by ibd1279** [comment]
Cannot say I am as lucky:
git clean -xdf && git reset --hard && git fetch origin && git pull origin bfgminer && ./autogen.sh && ./configure --enable-cpumining --enable-scrypt PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig:/usr/local/opt/jansson/lib/pkgconfig && make
...
sha256_xmm_amd64.asm:317: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.

using yasm 1.2; what version should I be using?

**Comment by luke-jr** [comment]
@ibd1279 Does this patch help? http://codepad.org/p6fXPB9P

**Comment by ibd1279** [comment]
That gets me a little further, but then I get this one:
sha256_sse4_amd64.asm:239: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.

I tried adding the qword onto what I think is the problem on that one (g_4sha256_k -- yasm error messages seem especially unhelpful), but then I get:
sha256_sse4_amd64.asm:239: error: invalid size for operand 2

**Comment by luke-jr** [comment]
Are you using the latest git? Line 239 doesn't have any code in it...

**Comment by ibd1279** [comment]
https://github.com/luke-jr/bfgminer/blob/bfgminer/x86_64/sha256_sse4_amd64.asm#L239 ??

239 is the unroll. That is what I meant by the yasm error messages are not very helpful. Anything that happens in that macro gets spit out where the unroll happens. That is why I am not sure if I was putting the qword even in the right place.

**Comment by luke-jr** [comment]
Ah, nasty. I suppose you could manually unroll the loop (even just once should expose the error)?

**Comment by ibd1279** [comment]
sha256_sse4_amd64.asm:237: error: macho: sorry, cannot apply 32 bit absolute relocations in 64 bit mode, consider "[_symbol wrt rip]" for mem access, "qword" and "dq _foo" for pointers.

line 237 is: movntdqa        xmm1, g_4sha256_k[rax*4]

sha256_sse4_amd64.asm:237: error: invalid size for operand 2
line 237 is: movntdqa        xmm1, qword g_4sha256_k[rax*4]

My assembly-fu is somewhere between "non-existant" and "I think I did a mov ax, bx thing once at band camp", but is this expected to use C-style pointer math? as in ((void*)((char*)g_4sha256_k) + (rax*4)) or is it supposed to some weird assembly/pointer math thing like ((void*)((char*)g_4sha256_k) + *((qword*)(rax*4)). I always thought square brackets meant "fetch the value from this address", and I don't remember identifier[offset] syntax... but my assembly-fu just plain sucks.

Ugh, escaping the dereference/pointer op sucks.

**Comment by luke-jr** [comment]
Try this: http://codepad.org/TyaNGJIn

**Comment by ibd1279** [comment]
That builds, but it doesn't love it running:
 [2013-05-31 00:51:24] "sse2_64"  : benchmarking algorithm ...Segmentation fault: 11
when I specify -a sse4_64, I also get a segfault: 11.

**Comment by fabulouspanda** [comment]
Sorry to have misled you before - I'd compiled without yasm so it didn't kick up a fuss.

**Comment by fabulouspanda** [comment]
I was able to build bfgminer a few days ago from git with ./configure PKG_CONFIG_PATH=/Applications/MacMiner.app/Contents/Resources/curl/lib/pkgconfig:/Applications/MacMiner.app/Contents/Resources/jansson/lib/pkgconfig --enable-scrypt --enable-cpumining CPPFLAGS="-I/uthash-master/src/"
(without yasm) - but now i'm getting segfault 11 from compiled git or 3.1
I tried compiling without cpu, without avalon but no luck.

there were just a few errors during compilation, mostly unused variables, and 
driver-avalon.c:939: warning: format ‘%06ld’ expects type ‘long int’, but argument 6 has type ‘__darwin_suseconds_t’

/usr/bin/ranlib: file: libgnu.a(dummy.o) has no symbols
ranlib: file: libgnu.a(dummy.o) has no symbols

**Comment by luke-jr** [comment]
@fabulouspanda  Can you bisect this?

**Comment by fabulouspanda** [comment]
It was this commit that broke it:
https://github.com/luke-jr/bfgminer/commit/564fa6ecb198e1e5649705459d804bab0f0e6352

**Comment by luke-jr** [comment]
Not if you compile without CPU... O.o

**Comment by fabulouspanda** [comment]
I was compiling without CPU. Go figure.

**Comment by fabulouspanda** [comment]
hang on, it seems like it works if I specifically state --disable-cpu although it says it's disabled after configure either way let me make sure i've not done something wrong here

**Comment by fabulouspanda** [comment]
Sorry I was running the wrong binary there probably. I just installed from git and it looks as though one of your recent commits has fixed it. Thanks!

**Comment by khera** [comment]
the 3.1.0 released tarball compile crashes with segfault every time I run it. Traced it down to this:

   Program received signal EXC_BAD_ACCESS, Could not access memory.
   Reason: KERN_INVALID_ADDRESS at address: 0x0000000078c37108
   0x00007fff8d7902f7 in flockfile ()
   (gdb) where
   #0  0x00007fff8d7902f7 in flockfile ()
   #1  0x00007fff8d792b18 in fseek ()
   #2  0x0000000100031a1b in file_contents (filename=0x7fff5fbfeb40 "poclbm130302.cl",  length=0x7fff5fbfeaf4) at ocl.c:245
   #3  0x000000010003275d in initCl (gpu=1606414192, name=0x7fff5fbfef70 "��_�", nameSize=140734799802224) at ocl.c:709
   #4  0x000000010002f098 in opencl_thread_prepare (thr=0x0) at driver-opencl.c:1595
   #5  0x00000001000170da in main (argc=1, argv=0x7fff5fbff9c0) at miner.c:9074
   (gdb) quit

However, if I build from the git HEAD, it seemingly works fine.

I use brew to install all the dependencies and then configured with this:

   env LIBCURL_LIBS="-L/usr/local/Cellar/curl/7.30.0/lib -lcurl" LIBCURL_CFLAGS=-I/usr/local/Cellar/curl/7.30.0/include ./configure

For what its worth, I have shared a brew for this, which is currently at 3.0.2 since 3.1.0 will not run. You can access it with "brew tap khera/homebrew-vick" then simply 'brew install bfgminer --HEAD' to get the latest from git. I try to keep it updated so normal brew upgrade should keep you up to date.

**Comment by caktux** [comment]
Getting `Unable to open poclbm130302.cl or ./poclbm130302.cl for reading`, I think open_bitstream() needs some attention.

**Comment by fabulouspanda** [comment]
I'm having trouble making 3.1.2

`In file included from fpgautils.c:775:
iospeeds.h: In function ‘tiospeed_t’:
iospeeds.h:22: error: ‘B460800’ undeclared (first use in this function)
iospeeds.h:22: error: (Each undeclared identifier is reported only once
iospeeds.h:22: error: for each function it appears in.)
iospeeds.h:23: error: ‘B921600’ undeclared (first use in this function)
iospeeds.h:25: error: ‘B576000’ undeclared (first use in this function)
iospeeds.h:26: error: ‘B1152000’ undeclared (first use in this function)
iospeeds.h:28: error: ‘B1500000’ undeclared (first use in this function)
iospeeds.h:29: error: ‘B3000000’ undeclared (first use in this function)
iospeeds.h:31: error: ‘B500000’ undeclared (first use in this function)
iospeeds.h:32: error: ‘B1000000’ undeclared (first use in this function)
iospeeds.h:33: error: ‘B2000000’ undeclared (first use in this function)
iospeeds.h:34: error: ‘B4000000’ undeclared (first use in this function)
iospeeds.h:36: error: ‘B2500000’ undeclared (first use in this function)
iospeeds.h:37: error: ‘B3500000’ undeclared (first use in this function)
make[2]: *** [bfgminer-fpgautils.o] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2`

**Comment by ibd1279** [comment]
I was having the same problem as fabulouspanda, but it appears fixed in the latest pull.

**Comment by sakkaku** [comment]
I got my jalapeño in today.  ~~Only caveat is you need to manually specify the device(s) when starting.~~

You may also need to install the [x64 drivers from here](http://www.ftdichip.com/Drivers/VCP.htm).  I am not 100% on this though.

Seems to be getting fairly fast speeds while running cool.

![BFGMiner on OSX w/ BFL](http://i.imgur.com/TlPN0uU.png)

**Comment by fabulouspanda** [comment]
You do need the ftdi usb serial drivers but the 3.1.3 version i pulled from github for the latest beta version of MacMiner (a few fixes from the released 3.1.3) allows you to use -S all rather than specifying the device, so if that doesn't work for you something's wrong in 3.1.4 - I haven't checked it yet though and not sure whether you tried that? I'm using a jala too.

**Comment by sakkaku** [comment]
`-S all` works.  Yay

**Comment by sakkaku** [comment]
I'm gonna go ahead and close the super mega OSX thread.

Good job to all the devs for keeping OSX a usable platform.
